File: Algorithms Techniques/2. Searching/binary-search/binarySearchIterative.py
# iterative binary search algorithm implementation
def binary_search(array, left, right, target):
  # check if the array is empty
  while left <= right:
    mid = left + (right - 1) / 2 # calculate the mid location
    if array[mid] == target: # if target is the mid value
      return mid
    if array[mid] > target: # iterate the left half
      right = mid - 1
    else: # iterate the right half
      left = mid + 1
  # can't find it in the array, return -1
  return -1

# testing the binary search algorithm
array = [1,2,3,4,5,6,7]
print("Array contents:")
print(array)

# search it
target_location = binary_search(array, 0, len(array)-1, 6)
if target_location == -1:
  print("Element 6 is not in the array")
else:
  print("Element 6 is at index " + str(target_location))

File: Algorithms Techniques/2. Searching/binary-search/binarySearchIterative.cpp
// how to run: c++ binarySearchIterative.cpp -std=c++11
#include <iostream>
#include <vector>
using namespace std;

// iterative approach
int binarySearch(vector<int>& array, int left, int right, int target) {
  // check if the array is empty
  while (left <= right) {
    int mid = left + (right - 1) / 2; // calculate the mid element
    if (array[mid] == target) // found it in the middle
      return mid;
    if (array[mid] > target) // iterate the left half
      right = mid - 1;
    else // iterate the right half
      left = mid + 1;
  }
  // can't find it, return -1
  return -1;
}

int main() {
  cout << "\nIterative Binary Search O(logn) Algorithm:\n";
  vector<int> array{2,4,5,6,7,8};
  cout << "Array indexes:  0 1 2 3 4 5\n";
  cout << "Array contents: ";
  for (int item : array)
    cout << item << " ";
  cout << endl;

  // run binary search to find the element
  int index = binarySearch(array, 0, array.size()-1, 7);
  if (index == -1)
    cout << "Element 7 not present in the array!" << endl;
  else 
    cout << "Found 7 at index " << index << endl;
}